src/main/scala/ir/dsl/WhileDSL.scala

```diff
@@ -0,0 +1,199 @@
+package ir.dsl
+import ir.*
+import translating.PrettyPrinter.*
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.collection.immutable.*
+import scala.annotation.targetName
+
+/**
+ * Construction of basil IR programs from a simple high level while langauge. 
+ *
+ * These functions uniformly return a list of eventually block, where the first block 
+ * is the entry point and the last block is the exit point. Generally this just means
+ * they meet the contract of [[sequence]] as either argument.
+ *
+ * [[sequence]] is used to join two lists of blocks such that control flows from the 
+ * first to the second.
+ *
+ * The function [[blocks]] folds [[sequence]] over a list of lists of blocks: 
+ *  essentially list.concat, but it connects each sub-list with gotos.
+ */
+
+object Counter {
+  var count = 1
+  def next() = {
+    count += 1
+    count
+  }
+
+  def nlabel(name: String = "id"): String = {
+    name + "_" + next()
+  }
+}
```

Comment:
consider using thread-safe counter in utils. it's _probably_ not a big issue here, but it would be nice to prevent counter creep